src/coreclr/jit/llvm.cpp

```diff
@@ -2387,6 +2427,55 @@ void Llvm::lowerCallToShadowStack(GenTreeCall* callNode)
     }
 }
 
+void Llvm::convertLclStructToLoad(GenTreeLclVarCommon* lclNode, ClassLayout* clsLayout)
+{
+    GenTree* lclAddr = _compiler->gtNewLclVarAddrNode(lclNode->GetLclNum());
+
+    lclNode->ChangeOper(GT_OBJ);
+    lclNode->AsObj()->SetAddr(lclAddr);
+    lclNode->AsObj()->SetLayout(clsLayout);
+
+    _currentRange->InsertBefore(lclNode, lclAddr);
+}
+
+void Llvm::lowerStoreLcl(GenTreeLclVarCommon* storeLclNode)
+{
+    GenTree* dataOp = storeLclNode->gtGetOp1();
+    CORINFO_CLASS_HANDLE dataHandle = _compiler->gtGetStructHandleIfPresent(dataOp);
+    if (dataOp->OperIs(GT_IND))
+    {
+        // Special case: "gtGetStructHandleIfPresent" sometimes guesses the handle from
+        // field sequences, but we will always need to transform TYP_STRUCT INDs into OBJs.
+        dataHandle = NO_CLASS_HANDLE;
+    }
+
+    LclVarDsc* addrVarDsc = _compiler->lvaGetDesc(storeLclNode->GetLclNum());
```

Comment:
This should be moved under the `storeLclNode->TypeIs(TYP_STRUCT)` clause, since it's not relevant for non-struct stores:

```c
if (storeLclNode->OperIsCopyBlkOp())
{
    dataHandle = { ... };
    if (addrVarDsc->GetStructHnd() != dataHandle) { ... }
}
```

Reply:
Are we removing the `if (dataOp->OperIs(GT_IND))` check ?

Reply:
I tried changing `storeLclNode->TypeIs(TYP_STRUCT)` to `storeLclNode->OperIsCopyBlkOp()` but I'm getting some assert.  Let me see what that is.

Reply:
Ah, I am sorry for that, `OperIsCopyBlkOp` will indeed not work here, it doesn't recognize `STORE_LCL_VAR`. `TypeIs(TYP_STRUCT)` should be left as it was.

(There is an interesting inconsistency here: `ASG(LCL struct, ...)` is `OperIsCopyBlkOp` but the equivalent `STORE_LCL_VAR` is)